[WIP] Earth (MARS)

jcrudy opened the pull request:
This pull request adds multivariate adaptive regression splines to scikit-learn.  
1. Adds the sklearn.earth module, which provides the Earth class
2. Adds an earth page to the documentation
3. Adds some earth examples

It probably could still use more work, but I'm looking for feedback on how best to proceed.  Currently my plan is to focus on documentation going forward.

jcrudy committed:
Made changes to pipeline to allow more general parameter passing

jcrudy committed:
Revert "Made changes to pipeline to allow more general parameter passing"

This reverts commit 87db631598a87003014513560bcc68161a7ca216.

jcrudy committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn

Getting recent changes from scikit-learn people.

jcrudy committed:
Merge remote-tracking branch 'upstream/master'

jcrudy committed:
Added earth.  Not passing all tests yet.

jcrudy committed:
Added documentation.  Still need to figure out how to deal with citations.

jcrudy committed:
Working on examples

jcrudy committed:
Updated the earth code from latest pyearth

jcrudy committed:
Merged upstream changes

jcrudy committed:
Fixe imports and got tests working

jcrudy committed:
Fixed imports again.  Not really sure how they got un-fixed.

I thought they were working before.  Maybe I ran the wrong tests or
something.

jcrudy committed:
Made the tests use relative imports

jcrudy committed:
Made Earth pass the common tests

jcrudy committed:
merged upstream

jcrudy committed:
removed unused gcv

jcrudy committed:
autopep8 on everything

jcrudy committed:
Fixed white space

jcrudy committed:
Merged classifier comparison and removed benchmark vs the R package

jcrudy committed:
Changed name of Earth to EarthRegressor

jcrudy committed:
Fixed the remaining examples to be more consistent with the rest of sklearn and to work with gen_rst

jcrudy committed:
Fixed titles of examples

jcrudy committed:
Changed doc page so that example is pulled from plot_v_functio.py

jcrudy committed:
Incorporated some new features and bug fixes from py-earth

1. Added allow_linear argument
2. Fixed a bug in the way endspan argument was processed
3. Fixed a bug in the way sample_weights were being applied
4. Added simple test case for 3 (test needs to be improved, may be
platform dependent)
5. Fixed a typo where min_search_points was accidentally
min_searh_points

ragv commented:
I feel it would be nice to have a todo list for this PR  :)

amueller commented:
I think someone needs starting to review to create a todo list. I am currently working on bugfixes for 0.16 (ugh sounds too much like the start of the discussion only two years later :-/) but I have quite some sklearn cycles at the moment, so maybe after that....

ragv commented:
> so maybe after that....

Sure :) I'd like to help... ( after the todo is framed )

mehdidc commented:
Hello there, I want to work on this PR and I can do the todo list.  I think the first step should be to fix the merge conflicts, no ?

agramfort commented:
to take over you should create a new PR with these commits and start addressing the issues. Style, examples, docs et perf benchmarks

jcrudy commented:
@mehdidc @ragv @amueller I am still happy to be involved, too, once there is a todo list.  My time is limited right now, but at least I can be an extra set of eyes and point out how things work where it isn't clear.  I would say that anyone working on this or taking it over should look at py-earth (https://github.com/jcrudy/py-earth), as there have been some improvements and bug fixes since I made this PR.

mehdidc commented:
great ! @jcrudy I was preparing a new PR based on the earth branch of your scikit learn fork, but it seems they are not synced (as you said), so now if I want to add a new PR, I start from py-earth, update the files that are not synced, then create the PR, right ?

ragv commented:
Also please move `earth.py` into `sklearn/additive` as discussed in #3482 in your new pr :)

jcrudy commented:
@mehdidc Yes, that's right.  I would ignore my scikit-learn fork, as it is way out of date.  Probably best to create your own fork of scikit-learn, make an earth branch, and add in the py-earth code under sklearn/additive as mentioned by @ragv .  That way there shouldn't be so much work to do getting everything synced up

Also, I've been planning to merge the derivatives branch of py-earth into master for some time, but have not yet.  It implements smoothing and first derivatives for MARS models, as described in Friedman's 1991 paper.  Basically, it let's you calculate the gradient of your model.  I'm not sure whether that's something that should go into sklearn, but if it is then that branch needs to be merged.  Maybe @amueller @agramfort @ragv would have an opinion about whether that feature should be included?

amueller commented:
I don't have a good feeling on what these models are used for, so I don't have a strong opinion. What kind of datasets do people use pyearth on?
What kind of benefits do smoothing and first derivatives bring you?

jcrudy commented:
@amueller For me they have been useful for figuring out the sensitivity of my model's output to small changes in its inputs.  Such knowledge sometimes has business uses for me.  In my case, I am usually modeling healthcare costs or risk of hospitalization.  I had one py-earth user (other than myself) request derivatives as a feature, but I don't know what he wanted them for.  Friedman does not offer much justification for them in his paper, but does go to the trouble of describing how to accomplish them.  Here is what he does offer, from section 3.7 of Friedman, 1991:

"If the intent is accurate estimation of the function (as opposed to its derivatives of various orders),  there is little to be gained by imposing continuity beyond that of the function itself.  If the true underlying function nowhere has a very large local second derivative, then a small additional increase in accuracy can be achieved by imposing continuous first derivatives.  Also, continuous (first) derivative approximations have considerable more cosmetic appeal."

However, smoothing and derivatives are not a feature of the R package earth, nor any other implementation that I know of, so are probably not commonly used.

amueller commented:
Thanks for the explanation. That sounds like we would probably not include them for the time being. On the other hand, it might be awkward if you include them in pyearth and we don't include them. I understood your intention was to not have pyearth as a separate project at some future point.

jcrudy commented:
@amueller Yes, that's true.  However, I have found maintenance of py-earth to not be much of a burden, and I'm sure even less so once it is in sklearn, so I am not too worried about that.  I suppose it is possible that over time the two could diverge in other ways, too, however, which might make life more complicated.  I guess my slight preference would be for inclusion, but for mostly reasons having to do with my own uses and ease of maintenance.

mblondel commented:
py-earth seems like a well-maintained project. I am not completely sure we need to merge it in scikit-learn.

agramfort commented:
how much is py-earth compliant with sklearn API / conventions?

would it make sense first to work in the py-earth repo to make it ready for
inclusion after showing the benefit on some datasets?

jcrudy commented:
@agramfort It is compliant as far as I know how to make it, or at least it was when I wrote it.  I expect there are probably some ways in which I have not followed conventions, but I am not aware of them.  I agree that working to make py-earth more compliant before trying to integrate it makes sense.  It would require someone more knowledgable about scikit-learn conventions than myself to give it a look and make a todo list.  I expect it would be a fairly short list, and I would happily accept a PR from @mehdidc or anyone else to address any issues.  I guess the issues are basically the same whichever order things are done in (fix and integrate or integrate and fix), so it really matters what works best for @mehdidc or whoever wants to push things forward.

@mblondel As far as leaving it as potentially a separate project, some downsides are:

1. Users of scikit-learn who want to use MARS might not be aware of py-earth
2. Py-earth requires a separate installation
3. It might create issues around maintaining compatibility

If the consensus is that py-earth is better as a separate project, I can put a little more effort into making it known and available.  I can at least put it on pypi.  I would also still welcome a todo list of compatibility issues and any pull requests addressing those issues.

amueller commented:
One of the things we want to have soon is a simple test that you can import and run it on your estimators to check compatibility. Not there yet, though :-/

jcrudy commented:
@amueller That would make things a lot easier as far as maintaining compatibility.

agramfort commented:
a quick look suggest there is a bit of work in py-earth to make sklearn
happy.

I would start with flake8 on all files, avoid public methods besides fit,
predict, transform, + checking import orders, starting examples with
plot_*, test coverage to reach > 90% as we have now in sklearn

jcrudy commented:
Great, thank you @agramfort.   A clarification: what do you mean by checking import orders?  @mehdidc, is there any part of that you're interested in taking on? 

As one of these involves removing some functionality (avoiding public methods other than fit, predict, and transform), I would suggest that at least that change should be done as part of integration into sklearn.  The rest can be done directly in py-earth.

mehdidc commented:
@jcrudy : In addition to what He suggested above, @agramfort  gave a good todo list here : #4341, I am working on it.

dsquareindia commented:
Hey everyone, I am planning to take this project of integrating pyearth into sklearn forward. I'll create a new branch, make a new `sklearn/additive`, pull all the py-earth code in it and work on that. But on the other hand, to integrate into sklearn later I'll have to merge into the files such as classes.rst so I was thinking of manually adding into these files and making incremental commits so that the review process can be faster. Should I go ahead with the second method?

mblondel commented:
We recently created a new organization called scikit-learn-contrib whose goal is to host scikit-learn compatible projects:
https://github.com/scikit-learn-contrib

This is very much a work in progress but our goal is to gather nice projects in the scikit-learn ecosystem. We just created it so only lightning is there but more projects will be added in the future.

We are thinking py-earth could possibly be moved there. The advantage is that py-earth would become more collaborative. When lightning was hosted in my account I was a bottleneck since I was the only one who could merge PRs. But now others have joined and bugs can be fixed more quickly.

Earth / Mars is sufficiently mature to warrant inclusion in scikit-learn but moving it to scikit-learn-contrib would probably involve much less work.

agramfort commented:
cc @jcrudy

dsquareindia commented:
So probably one of the first things we should do is add a description of this in the scikit learn readme so that the users get to know about it.

agramfort commented:
it's very fresh. We'll do it when we really advertise the
scikit-learn-contrib org.

mblondel commented:
Yes it's too early to advertise it. I just informed you so you don't loose time.

dsquareindia commented:
@agramfort just wanted to ask one more thing, are the remaining projects of adding GAM (and for LSS), SpAM and LISO to sklearn still of any interest?

agramfort commented:
I don't consider them as priorities; The > 400 open PRs we have are a
bigger concern to me to be honest...

mblondel commented:
@jcrudy This is very much a work in progress but I have uploaded the basic workflow for inclusion in scikit-learn-contrib. If you're interested in moving py-earth there, we can provide some assistance.

jcrudy commented:
@mblondel Yes, I would like to do that.  I think it's a great idea.  I'm probably too busy to look at it much for the next couple weeks, but I will as soon as I can, and I'll start with the workflow you uploaded.  Is there anything else I should keep in mind?

agramfort commented:
@mehdidc are you up for it?

mehdidc commented:
@agramfort Yes !  @jcrudy  @mblondel  Ok so I first go through the requirements  then add a request in scikit-learn-contrib

jcrudy commented:
@mehdidc There is a branch of py-earth called `rewrite_forward_pass` that I've been working on.  I'm planning to merge it into master in the not too distant future, but it isn't quite ready yet.  It includes some bug fixes to missing data functionality and allows arbitrary (non-separable) sample weights.  It also refactors/rewrites the forward pass in a way that is less messy, and uses householder reflections instead of gram-schmidt for orthogonalization of the `B` matrix (the main reason I did it, as I was having stability problems on a project).  Just want you to be aware so you don't repeat work.  If you find major changes are needed to pass the contrib requirements, let's talk before you do them.  (Because it might be better to merge first.)  Also, let me know if I can answer any questions or be of assistance in some other way.

mehdidc commented:
@jcrudy Great ! it will be nice to have arbitrary sample weights. Ok  I see, thanks I will let you know

springcoil commented:
I see that @jcrudy made some progress with Py-Earth. Which I think make it even more appetising. I know there are lots of open PR's for Scikitlearn. I've made a few reviews of code - since that's all I can offer the project at the moment.

mblondel commented:
If it's ok with @jcrudy and @mehdidc, I think we should transfer [*] py-earth to the contrib organization and collaboratively work on meeting the requirements there. We need a project base so as to make the contrib more attractive. Currently, with lighting only, this feels a bit sad.

[*] Transferring rather than forking is important so as to not loose branches and stars.

jcrudy commented:
@mblondel That's fine with me.  @mehdidc, will that mess up anything you're doing?

mehdidc commented:
@mblondel @jcrudy  No, it's fine with me too

jcrudy commented:
@mblondel  @mehdidc Okay, I'll transfer it over.  Exciting times!

jcrudy commented:
@mblondel Do I need to be added to the contrib organization in order to make the transfer?

https://help.github.com/articles/transferring-a-repository/#transferring-from-a-user-to-an-organization

agramfort commented:
@jcrudy you're now a member of scikit-learn-contrib org :)

jcrudy commented:
@agramfort @mehdidc @mblondel Transfer complete.  The py-earth project is now a part of scikit-learn-contrib!

How is progress on making it actually compliant with the requirements?  It seems it might be actually there, since @mehdidc has addressed the problems it was having with `check_estimator`.  If that is complete, I think the next step is to work toward a release.

mblondel commented:
@jcrudy Thanks a lot! You might also want to fork the project in your github. This way, the old project URL is still valid.

mblondel commented:
Closed this pull-request. Let's create an issue directly in the project to follow the progress of meeting the requirements.

mblondel commented:
I added a py-earth team. @jcrudy you should be able to add collaborators to the team.
